Only used PyUnicode_Concat on unicode objects

da-woods opened the pull request:
Closes https://github.com/cython/cython/issues/3426

-------------------------

~(I'm getting a segmentation fault on the fstring tests with the current master, which is making it difficult to run this through a thorough set of tests. It happens with and without this PR. I'm not sure where this is due to something being messed up on my side or a real bug, but I'll look into and report it separately. Just noting it here because it means I couldn't test this PR with with "runtest.py string")~ Caching issues to do with Cython.inline - I'd clearly ended up with a broken version cached - ignore

scoder commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -11364,13 +11364,8 @@ def compute_c_result_type(self, type1, type2):
-            if type1 is unicode_type or type2 is unicode_type:
-                is_unicode_concat = type1.is_builtin_type and type2.is_builtin_type
+        type1, type2 = self.operand1.type, self.operand2.type
+        is_unicode_concat = type1 is unicode_type and type2 is unicode_type
```
Looks like the right place to fix this, but I think we should also allow `str_type` here if one side is `unicode_type`.

da-woods replied:
bytes + unicode doesn't work in Python - try `u"ab" + b"cd"` and `b"cd" + u"ab"`.

da-woods replied:
However... there is one optimization it might actually be worth looking into here (although only tangentially related to this bug). CPython has a specific optimization to check if the reference count of the RHS and append rather than concat if possible: https://github.com/python/cpython/blob/309d7cc5df4e2bf3086c49eb2b1b56b929554500/Python/ceval.c#L5354

scoder replied:
> bytes + unicode doesn't work in Python 3

Note: I wrote `str_type`, not `bytes_type`. They are different things in Cython.

scoder commented on tests/run/test_unicode_string_tests.pxi:
```diff
@@ -993,6 +993,13 @@ class CommonTest(BaseTest):
+        y = []
+        y += 'ab'
+        self.assertEqual('a', y[0])
+        self.assertEqual('b', y[1])
```
```suggestion
        self.assertEqual('a', y[0])
        self.assertEqual('b', y[1])
        self.assertEqual(['a', 'b'], y)
```

da-woods commented:
I've updated this to also optimize `str_type` where its clear that it's unicode (i.e. when language_level = 3).

I've also added in some optimization of inplace concatenation. This does give a significant speed-up in some (probably mostly artificial) cases. However, it also uses some slightly awkward macros to make it fit. If you'd rather drop it then I can just remove https://github.com/cython/cython/pull/3433/commits/7d2608ecb96c6e4194490c6804368b41a2591ba7

scoder commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -11364,13 +11364,13 @@ def compute_c_result_type(self, type1, type2):
+        # in Py3 str_type is unicode so can be concatted to unicode_type or str_type
+        is_unicode_concat = (type1 in (unicode_type, str_type) and type2 in ((unicode_type, str_type))
+                                and not (type1 is str_type and type2 is str_type
+                                         and language_level == 2))
```
Is there a reason why we need to look at the `language_level` here?
ISTM that all we need is 1) one of `type1` and `type2` is `unicode_type`, and 2) the other is either `unicode_type` or `str_type`.

scoder commented:
> CPython has a specific optimization to check the reference count of the RHS and append rather than concat if possible

Yes, that seems worth doing on our side as well. Especially since 2-value f-strings are already replaced by a concatenation rather than joining because it's much faster. Making this even faster with this "hack" would be nice.

scoder commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -11364,13 +11364,8 @@ def compute_c_result_type(self, type1, type2):
     def py_operation_function(self, code):
-        is_unicode_concat = False
-        if isinstance(self.operand1, FormattedValueNode) or isinstance(self.operand2, FormattedValueNode):
-            is_unicode_concat = True
```
I wonder if removing this special case can lead to regressions. Doesn't seem very likely, given that f-strings are always built from unicode string parts. But I don't remember all the optimisations that apply to these parts along the way. Not all of them may be aware of the concrete result type in order to preserve it along the way.

da-woods replied:
As far as I could tell the result type of `FormattedValueNode` was permanently set to `unicode_type` anyway, so it seemed fairly safe. Also, it'd be (probably fairly small?) performance regression rather than than anything breaking (`PyNumber_Add` does work). I can definitely put it back in if wanted, but it seems more likely to cause bugs then fix them.

scoder replied:
> seems more likely to cause bugs then fix them.

`FormattedValueNode` should only originate from f-strings, and they always produce string parts that are then joined in one way or another. That seems safe. The question is just if the new "unicode or str" decision below succeeds for all of those parts, including operands that are no longer `FormattedValueNode`s themselves (but return a unicode string at runtime).

This PR seems like a fix that should go into 0.29.x, really, in which case I'd like to reduce the risk of regressions a bit more than strictly necessary. I don't think it hurts to keep the `isinstance()` tests, at least in that branch.

da-woods commented:
> > CPython has a specific optimization to check the reference count of the RHS and append rather than concat if possible
> 
> Yes, that seems worth doing on our side as well. Especially since 2-value f-strings are already replaced by a concatenation rather than joining because it's much faster. Making this even faster with this "hack" would be nice.

I've pushed my patch for this. I had managed to get it working but was a bit undecided about if it should go in this PR. Notes from the relevant commit below

---------------------------

Do unicode/bytes concatenation inplace if possible
    
* Slightly odd-macro trickery - does modify the input argument too
    (although only in places where it shouldn't matter - but potentially
    should  we worry about XDECREF?)
* Doesn't (yet) happen on str_type+str_type on language_level=2/3str
    because the type of str_type isn't known at the Cythonizing stage
* Doesn't (and probably can't) happen in cases where Cython hasn't
    identified the types correctly and is just using PyNumber_Add.
* For:
    
        try:
           a += "a"
        except:
           # stuff
    
   `a` will be NULL in stuff. The only exception we should be able
    to get is `MemoryError` and the CPython interpretter shows the same
    behaviour in principle (it's noted as an issue in `ceval` for Python2)

scoder commented on Cython/Utility/ObjectHandling.c:
```diff
@@ -2553,3 +2553,101 @@ static PyObject *__Pyx_PyMethod_New(PyObject *func, PyObject *self, CYTHON_UNUSE
+
+/////////////// UnicodeConcatInplace.proto ////////////////
+
+#if CYTHON_COMPILING_IN_PYPY || PY_MAJOR_VERSION == 2
```
This seems more of a CPython-only optimisation than a not-PyPy thing.
(And it feels worth a separate PR.)

da-woods replied:
OK - backed it out again. I'll submit it as a separate PR.
